src/librustc_trans/back/write.rs

```diff
@@ -1109,164 +1201,230 @@ fn execute_work<'a>(sess: &'a Session,
+                                .. cgcx.clone()
+                            };
+
+                            trans_worker_state = TransWorkerState::LLVMing;
+                            spawn_work(cgcx, item);
+                        }
+                    }
+                    TransWorkerState::Translating => {
+                        bug!("trans worker should not be translating after \
+                              translation was already completed")
+                    }
+                    TransWorkerState::LLVMing => {
+                        // Already making good use of that token
+                    }
+                }
             }
-            Message::Done { success: false } => {
-                sess.fatal("aborting due to worker thread panic");
+
+            // Spin up what work we can, only doing this while we've got available
+            // parallelism slots and work left to spawn.
+            while work_items.len() > 0 && running < tokens.len() {
+                let item = work_items.pop().unwrap();
+                let worker_id = get_worker_id(&mut free_worker_ids);
+
+                let cgcx = CodegenContext {
+                    worker: worker_id,
+                    .. cgcx.clone()
+                };
+
+                spawn_work(cgcx, item);
+                running += 1;
             }
 
-            // Our worker wants us to emit an error message, so get ahold of our
-            // `sess` and print it out
-            Message::Diagnostic(diag) => {
-                let handler = sess.diagnostic();
-                match diag.code {
-                    Some(ref code) => {
-                        handler.emit_with_code(&MultiSpan::new(),
-                                               &diag.msg,
-                                               &code,
-                                               diag.lvl);
+            // Relinquish accidentally acquired extra tokens
+            tokens.truncate(running);
+
+            match coordinator_receive.recv().unwrap() {
+                // Save the token locally and the next turn of the loop will use
+                // this to spawn a new unit of work, or it may get dropped
+                // immediately if we have no more work to spawn.
+                Message::Token(token) => {
+                    if let Ok(token) = token {
+                        tokens.push(token);
+                    } else {
+                        shared_emitter.fatal("failed to acquire jobserver token");
+                        // Exit the coordinator thread
+                        panic!()
                     }
-                    None => {
-                        handler.emit(&MultiSpan::new(),
-                                     &diag.msg,
-                                     diag.lvl);
+                }
+
+                Message::TranslationDone { llvm_work_item, is_last } => {
+                    work_items.insert(0, llvm_work_item);
```

Comment:
No reason. I changed that during a debugging session. I'll switch it back to a `push`.